// @shopify/polaris/build/cjs/components/DropZone/components/FileUpload/FileUpload.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var css = require('../../../../utilities/css.js');
var capitalize = require('../../../../utilities/capitalize.js');
var context = require('../../context.js');
var index = require('../../utils/index.js');
var FileUpload$1 = require('./FileUpload.scss.js');
var uploadArrow = require('../../images/upload-arrow.svg.js');
var Caption = require('../../../Caption/Caption.js');
var hooks = require('../../../../utilities/i18n/hooks.js');
var Stack = require('../../../Stack/Stack.js');
var TextStyle = require('../../../TextStyle/TextStyle.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function FileUpload(props) {
  const i18n = hooks.useI18n();
  const {
    size,
    measuring,
    type,
    focused,
    disabled,
    allowMultiple
  } = React.useContext(context.DropZoneContext);
  const typeSuffix = capitalize.capitalize(type);
  const allowMultipleKey = index.createAllowMultipleKey(allowMultiple);
  const {
    actionTitle = i18n.translate(`Polaris.DropZone.${allowMultipleKey}.actionTitle${typeSuffix}`),
    actionHint = i18n.translate(`Polaris.DropZone.${allowMultipleKey}.actionHint${typeSuffix}`)
  } = props;
  const buttonStyles = css.classNames(FileUpload$1['default'].Button, FileUpload$1['default'].slim, focused && FileUpload$1['default'].focused, disabled && FileUpload$1['default'].disabled);
  const buttonMarkup = size === 'large' && buttonStyles ? /*#__PURE__*/React__default['default'].createElement("div", {
    className: buttonStyles
  }, actionTitle) : null;
  const actionTitleClassName = css.classNames(FileUpload$1['default'].ActionTitle, focused && !disabled && FileUpload$1['default']['ActionTitle-focused'], disabled && FileUpload$1['default']['ActionTitle-disabled']);
  const actionTitleMarkup = /*#__PURE__*/React__default['default'].createElement("div", {
    className: actionTitleClassName
  }, actionTitle);
  const fileUploadClassName = css.classNames(FileUpload$1['default'].FileUpload, measuring && FileUpload$1['default'].measuring, size === 'large' && FileUpload$1['default'].large, size === 'small' && FileUpload$1['default'].small);
  let viewMarkup;

  switch (size) {
    case 'large':
      viewMarkup = /*#__PURE__*/React__default['default'].createElement(Stack.Stack, {
        vertical: true,
        spacing: "tight"
      }, /*#__PURE__*/React__default['default'].createElement("img", {
        width: "40",
        src: uploadArrow['default'],
        alt: ""
      }), buttonMarkup, /*#__PURE__*/React__default['default'].createElement(Caption.Caption, null, /*#__PURE__*/React__default['default'].createElement(TextStyle.TextStyle, {
        variation: "subdued"
      }, actionHint)));
      break;

    case 'medium':
      viewMarkup = /*#__PURE__*/React__default['default'].createElement(Stack.Stack, {
        vertical: true,
        spacing: "tight"
      }, actionTitleMarkup, /*#__PURE__*/React__default['default'].createElement(Caption.Caption, null, /*#__PURE__*/React__default['default'].createElement(TextStyle.TextStyle, {
        variation: "subdued"
      }, actionHint)));
      break;

    case 'small':
      viewMarkup = /*#__PURE__*/React__default['default'].createElement("img", {
        width: "20",
        src: uploadArrow['default'],
        alt: ""
      });
      break;
  }

  return /*#__PURE__*/React__default['default'].createElement("div", {
    className: fileUploadClassName
  }, viewMarkup);
}

exports.FileUpload = FileUpload;
