// @shopify/polaris/build/cjs/components/ResourceList/components/FilterControl/components/FilterCreator/FilterCreator.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var useToggle = require('../../../../../../utilities/use-toggle.js');
var Form = require('../../../../../Form/Form.js');
var FormLayout = require('../../../../../FormLayout/FormLayout.js');
var FilterValueSelector = require('../FilterValueSelector/FilterValueSelector.js');
var hooks = require('../../../../../../utilities/i18n/hooks.js');
var Button = require('../../../../../Button/Button.js');
var Popover = require('../../../../../Popover/Popover.js');
var Select = require('../../../../../Select/Select.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function FilterCreator({
  filters,
  resourceName,
  disabled,
  onAddFilter
}) {
  const {
    value: popoverActive,
    toggle: togglePopoverActive,
    setFalse: setPopoverActiveFalse
  } = useToggle.useToggle(false);
  const [selectedFilter, setSelectedFilter] = React.useState();
  const [selectedFilterKey, setSelectedFilterKey] = React.useState();
  const [selectedFilterValue, setSelectedFilterValue] = React.useState();
  const i18n = hooks.useI18n();
  const node = React.useRef(null);
  const canAddFilter = Boolean(selectedFilter && selectedFilterKey && selectedFilterValue);
  const handleButtonFocus = React.useCallback((...args) => {
    const event = args[0];

    if (!node.current && event) {
      node.current = event.target;
    }
  }, []);
  const handleFilterKeyChange = React.useCallback(filterKey => {
    const foundFilter = filters.find(filter => {
      const {
        minKey,
        maxKey,
        operatorText
      } = filter;

      if (minKey || maxKey) {
        return filter.key === filterKey || minKey === filterKey || maxKey === filterKey;
      }

      if (operatorText && typeof operatorText !== 'string') {
        return filter.key === filterKey || operatorText.filter(({
          key
        }) => key === filterKey).length === 1;
      }

      return filter.key === filterKey;
    });

    if (!foundFilter) {
      return;
    }

    setSelectedFilter(foundFilter);
    setSelectedFilterKey(filterKey);
    setSelectedFilterValue(undefined);
  }, [filters]);
  const handleFilterValueChange = React.useCallback(value => {
    setSelectedFilterValue(value);
  }, []);
  const handleAddFilter = React.useCallback(() => {
    if (!onAddFilter || !canAddFilter || !selectedFilterKey) {
      return;
    }

    onAddFilter({
      key: selectedFilterKey,
      value: selectedFilterValue || ''
    });
    setPopoverActiveFalse();
    setSelectedFilter(undefined);
    setSelectedFilterValue(undefined);

    if (node.current != null) {
      node.current.focus();
    }
  }, [canAddFilter, onAddFilter, selectedFilterKey, selectedFilterValue, setPopoverActiveFalse]);
  const activator = /*#__PURE__*/React__default['default'].createElement(Button.Button, {
    onClick: togglePopoverActive,
    disclosure: true,
    disabled: disabled,
    onFocus: handleButtonFocus
  }, i18n.translate('Polaris.ResourceList.FilterCreator.filterButtonLabel'));
  const filterOptions = filters.map(({
    key,
    label
  }) => ({
    value: key,
    label
  }));
  const filterValueSelectionMarkup = selectedFilter ? /*#__PURE__*/React__default['default'].createElement(FilterValueSelector.FilterValueSelector, {
    filter: selectedFilter,
    filterKey: selectedFilterKey,
    value: selectedFilterValue,
    onFilterKeyChange: handleFilterKeyChange,
    onChange: handleFilterValueChange
  }) : null;
  const addFilterButtonMarkup = selectedFilter ? /*#__PURE__*/React__default['default'].createElement(Button.Button, {
    onClick: handleAddFilter,
    disabled: !canAddFilter
  }, i18n.translate('Polaris.ResourceList.FilterCreator.addFilterButtonLabel')) : null;
  return /*#__PURE__*/React__default['default'].createElement(Popover.Popover, {
    active: popoverActive,
    activator: activator,
    onClose: togglePopoverActive,
    sectioned: true,
    fullHeight: true
  }, /*#__PURE__*/React__default['default'].createElement(Form.Form, {
    onSubmit: handleAddFilter
  }, /*#__PURE__*/React__default['default'].createElement(FormLayout.FormLayout, null, /*#__PURE__*/React__default['default'].createElement(Select.Select, {
    label: i18n.translate('Polaris.ResourceList.FilterCreator.showAllWhere', {
      resourceNamePlural: resourceName.plural.toLocaleLowerCase()
    }),
    placeholder: i18n.translate('Polaris.ResourceList.FilterCreator.selectFilterKeyPlaceholder'),
    options: filterOptions,
    onChange: handleFilterKeyChange,
    value: selectedFilter && selectedFilter.key
  }), filterValueSelectionMarkup, addFilterButtonMarkup)));
}

exports.FilterCreator = FilterCreator;
